lib/matplotlib/dates.py

```diff
@@ -201,6 +220,71 @@ def _get_rc_timezone():
     MO, TU, WE, TH, FR, SA, SU)
 WEEKDAYS = (MONDAY, TUESDAY, WEDNESDAY, THURSDAY, FRIDAY, SATURDAY, SUNDAY)
 
+# default epoch: passed to np.datetime64...
+_epoch = None
+_epoch_used = False
+
+
+def _reset_epoch_test_example():
+    """Reset the Matplotlib date epoch to its rcParams value."""
+    global _epoch
+    global _epoch_used
+
+    _epoch_used = False
+    _epoch = matplotlib.rcParams['date.epoch']
+
+
+def set_epoch(epoch):
+    """
+    Set the epoch (origin for dates) for datetime calculations.
+
+    Parameters
+    ----------
+    epoch : str
+        valid UTC date parsable by `numpy.datetime64` (do not include
+        timezone).
+
+    The default epoch is :rc:`dates.epoch`, and users should not need to
```

Comment:
This made me think: Do we need `rcParams['date.epoch']` at all if users should not modify it? Wouldn't be a hard-coded constant be enough and also prevent users from changing it in matplotlibrc or style files, which IMHO they shouldn't anyway.

At first I thought, you need the rcParam for a context, but that wouldn't work either because we don't allow different values throughout the execution.

OTOH that again made me think: Can we really globally deny changing the epoch. That means during program execution, we don't allow two different plots with different epochs.

Reply:
The concern is folks who used the old epoch and don't want to change their code, hence the rcParam.  

Its also possible that someone wants nanoseconds for a date this year, and wants to set the epoch to something very close to get down to nanoseconds.  

Indeed we do not allow two different epochs during program execution.  The reason was use cases like the one discussed by @ImportanceOfBeingErnest (https://github.com/matplotlib/matplotlib/pull/15008#issuecomment-519756008).  

I think the way things are now is a good balance of protective versus flexible, but happy to reconsider.

Reply:
> The concern is folks who used the old epoch and don't want to change their code, hence the rcParam.

Then I think it‘s still clearer and less dangerous to put `set_epoch()` directly after your matplotlib Import rather than putting the epoch in a `matplotlibrc` file. The file will most likely apply to more code and you can easily forget that it‘s there.

Anyway if you still want to keep the parameter, I‘m a bit reluctant to add a config value with the general statement that „users should not need to change this“. That doesn’t make sense. If you want to keep the parameter, write „users should usually not need to change this“. Or if you want to be precise, state when the parameter might be changed.

Reply:
To me, the point of having a matplotlibrc is to allow a user to set defaults.  I realize there is also the style sheet way of thinking of rcParams, but I think that should have been separate for things that we think can be mutable.  

In this case `date.epoch` is definitely meant in the immutable form.  I can readily imagine there are users who are perfectly happy to keep the old epoch, and want nothing to do with our changing the default, and they will be well-served by setting this in the matplotlibrc.

Reply:
Changed to 
`The default epoch is :rc:`dates.epoch` (by default 1970-01-01T00:00).`

Reply:
@timhoffm any further thoughts or is this OK?